src/egl/drivers/dri2/platform_android.c

```diff
@@ -669,11 +669,18 @@ droid_query_buffer_age(_EGLDriver *drv,
 {
    struct dri2_egl_surface *dri2_surf = dri2_egl_surface(surface);
 
+   /* To avoid blocking other EGL calls, release the display mutex before
+    * we enter droid_window_dequeue_buffer() and re-acquire the mutex upon
+    * return.
+    */
+   mtx_unlock(&disp->Mutex);
```

Comment:
@tpalli  Any comments ?

Reply:
@strassek ping also

Reply:
Did you try to move this down below to update_buffers, or droid_window_dequeue_buffer? It's risky to unlock the mutex as now some other function could be called at the same time. Is it actually happening this case, that releasing the mutex gives something else possibility to execute?

Reply:
It's to prevent the deadlock situation happen in android:
1) Consumer thread: updateTexImage() -> updateAndReleaseLocked() -> syncForReleaseLocked() -> eglDupNativeFenceFDANDROID() -> _eglLockDisplay() -> waiting for the dpy lock
2) Producer thread: eglQuerySurface() (acquired dpy lock) -> droid_query_buffer_age() -> update_buffers() -> android::Surface::dequeueBuffer() -> android::BufferQueueProducer::waitForFreeSlotThenRelock()
If we don't release the dpy lock here, the consumer thread will blocked and there's no empty slot for the producer thread, which leads to a deadlock.
and the variable disp will not be used during the update_buffers().

There's already a similar patch in another function.
commit 1ea233c6f30a74e6ff5456c3521328237b01eed8
Author: Haixia Shi <hshi@chromium.org>
Date:   Thu Jun 2 12:48:23 2016 -0700

    platform_android: prevent deadlock in droid_swap_buffers

Reply:
Right, that makes sense, thanks for the explanation. I think it would be good to mention points 1 and 2 in the commit message. Thanks and please send to mesa-dev mailing list!

Reply:
@minhe1  @chenlin3 Will one of you send this patch to Mesa mailing list for review ?